src/lib.rs

```diff
@@ -84,3 +84,15 @@ unsafe impl GlobalAlloc for Mimalloc {
         ptr as *mut u8
     }
 }
+
+/// Returns the available bytes in the memory block, or 0 if `ptr` was NULL.
+///
+/// The returned size is always at least equal to the allocated size of `ptr`,
+/// and, in the current design, should be less than 16.7% more.
```

Comment:
Not sure I understand this, can you elaborate? Does it mean that the result of calling `usable_size` is overestimated by at most 16.7%?

Reply:
Yes, this is copied from the mimalloc docs (see the link above) and is an implementation detail. Other allocators can also overestimate `usable_size`.

Reply:
This means that usable_size might return a size that's 16.7% less than the actual allocation size. But always larger than the size used to request the allocation, right ?

Reply:
@gnzlbg mmmh, no that's not how I read it. "at least equal to the allocated size", so it never returns less than the actual size, but __may__ overestimate it. If it does overestimate, the error is at most 16.7%. Or am I reading it wrong?

Reply:
I understand this sentence the same way @dvdplm does, the `more` part at the end refers to overestimation.

Reply:
So what can you do with the result of this API ? You can't use the size returned, because it could be an overestimation.

Reply:
> The returned size can be used to call mi_expand successfully.

Also it can be used to (approx) measure heap allocations (like in https://github.com/servo/heapsize).

Reply:
Note that the POSIX usable_size API guarantees that you can use the returned size as if it were the size that you performed the allocation with.

Reply:
```c++
/// The returned size can be
/// used to call \a mi_expand successfully.
/// The returned size is always at least equal to the
/// allocated size of \a p, and, in the current design,
/// should be less than 16.7% more.
```

I read this as "the returned size os between 1.0x and 1.16x the original allocation size".

Reply:
@gnzlbg what about jemallocator?
> /// Return the usable size of the allocation pointed to by ptr.
> /// The return value may be larger than the size that was requested during allocation.
/// This function is not a mechanism for in-place `realloc()`;
/// rather it is provided solely as a tool for introspection purposes.
/// Any discrepancy between the requested allocation size
/// and the size reported by this function should not be depended on,
/// since such behavior is entirely implementation-dependent.

Reply:
That is, when you call malloc, it will at most give you a block that's 16% larger than what you requested. But `usable_size` will still return the accurate size of the block, and you can use that in full.

Reply:
That doesn't mean that the size of the allocation is overestimated. What this means is that the memory block where the allocation lives can be greater than the allocation, which is standard practice.

Reply:
Jemallocator provides the same exact same semantics that `malloc_usable_size` has in all platforms, e.g., says exactly that as well: http://man7.org/linux/man-pages/man3/malloc_usable_size.3.html

Reply:
The recommendation for using this also for debugging purposes is in glibc as well, because `mi_usable_size` is _super slow_ compared with other functions to query the size of an allocation.

Reply:
I see, thanks for the clarification. The confusion comes from using `the allocated size of 'ptr'` wording instead `requested size of allocation for 'ptr'`.